content/development/testing.md

```diff
@@ -9,7 +9,310 @@ chapter: false
+                proxy_connect_timeout 60s;
+                proxy_read_timeout 36000s;
+                proxy_redirect off;
+
+                proxy_pass_header Authorization;
+        }
+```
+
+In both cases (Apache httpd, nginx) you have to change your `modsecurity.conf` settings. Open that file and find the directive `SecResponseBodyMimeType`. Modify the arguments:
+
+```
+SecResponseBodyMimeType text/plain text/html text/xml application/json
+```
+
+Note, that the default value does not have the MIME type `application/json`.
+
+In your `crs-setup.conf` you need to add these extra rules (after the rule `900990`):
+
+```
+SecAction
+    "id:900005,\
+    phase:1,\
+    nolog,\
+    pass,\
+    ctl:ruleEngine=DetectionOnly,\
+    ctl:ruleRemoveById=910000,\
+    setvar:tx.blocking_paranoia_level=4,\
+    setvar:tx.crs_validate_utf8_encoding=1,\
+    setvar:tx.arg_name_length=100,\
+    setvar:tx.arg_length=400,\
+    setvar:tx.max_file_size=64100,\
+    setvar:tx.combined_file_sizes=65535"
+
+SecRule REQUEST_HEADERS:X-CRS-Test "@rx ^.*$" \
+    "id:999999,\
+    phase:1,\
+    pass,\
+    t:none,\
+    log,\
+    msg:'%{MATCHED_VAR}'"
+```
+
+Now, after the restarting the web server all request will be sent to `httpbin`. Let's start testing.
+
+## Go-ftw
+
+Tests are performed using [go-ftw](https://github.com/coreruleset/go-ftw). We run our test suite automatically using **go-ftw** as part of a [GitHub workflow](https://github.com/coreruleset/coreruleset/blob/{{< param crs_dev_branch >}}/.github/workflows/test.yml). You can easily reproduce that locally, on your workstation.
+
+For that you will need:
+
+- the CRS Git repository
+- Docker (modern versions of docker already include the `compose` command, if you are running an older version you also need to have `docker-compose` installed)
+  OR
+  your own environment (see above installation steps)
+- your rules and tests!
+
+You can download pre-compiled binaries of **go-ftw** or build from source (requires you to have a **Go** environment). The pre-compiled binaries are available [on GitHub](https://github.com/coreruleset/go-ftw/releases). The binaries are ready to run and do not require installation.
+
+You can also install pre-compiled binaries by using `go install`, if you have a **Go** environment:
+
+```bash
+$ go install github.com/coreruleset/go-ftw@latest
+```
+
+This will install the binary into your `$HOME/go/bin` directory. To compile **go-ftw** from source, run the following commands:
+
+```bash
+$ git clone https://github.com/coreruleset/go-ftw.git
+$ cd go-ftw
+$ go build
+```
+
+This will build the binary in the **go-ftw** repository.
+
+Now create a configuration file. Because Apache httpd and nginx use different log file paths, and, perhaps, different ports, you may want to create two different configuration files for **go-ftw**. For details please read the [original](https://github.com/coreruleset/go-ftw#yaml-config-file) documentation of **fo-ftw**.
```

Comment:
```suggestion
Now create a configuration file. Because Apache httpd and nginx use different log file paths, and, perhaps, different ports, you may want to create two different configuration files for **go-ftw**. For details please read [go-ftw's documentation](https://github.com/coreruleset/go-ftw#yaml-config-file).
```